File: 01_duplicate_file_finder/java/Solution.java
import java.io.*;
import java.nio.file.*;
import java.security.*;
import java.util.*;

public class Solution {

    private static String getFileHash(Path path) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        try (InputStream is = Files.newInputStream(path)) {
            byte[] buffer = new byte[4096];
            int read;
            while ((read = is.read(buffer)) != -1) {
                md.update(buffer, 0, read);
            }
        }
        byte[] hashBytes = md.digest();
        StringBuilder sb = new StringBuilder();
        for (byte b : hashBytes)
            sb.append(String.format("%02x", b));
        return sb.toString();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter directory path: ");
        Path dir = Paths.get(sc.nextLine());

        Map<String, List<Path>> fileMap = new HashMap<>();

        try {
            Files.walk(dir)
                 .filter(Files::isRegularFile)
                 .forEach(path -> {
                     try {
                         String hash = getFileHash(path);
                         fileMap.computeIfAbsent(hash, k -> new ArrayList<>()).add(path);
                     } catch (Exception ignored) {}
                 });

            boolean found = false;
            for (List<Path> paths : fileMap.values()) {
                if (paths.size() > 1) {
                    found = true;
                    System.out.println("Duplicate group:");
                    paths.forEach(System.out::println);
                    System.out.println("-------------------------");
                }
            }

            if (!found)
                System.out.println("No duplicate files found.");

        } catch (IOException e) {
            System.out.println("Invalid directory.");
        }
    }
}

File: 01_duplicate_file_finder/python/solution.py
import os
import hashlib
from collections import defaultdict

def get_file_hash(file_path, chunk_size=4096):
    hasher = hashlib.md5()
    with open(file_path, 'rb') as file:
        while chunk := file.read(chunk_size):
            hasher.update(chunk)
    return hasher.hexdigest()

def find_duplicates(directory):
    hashes = defaultdict(list)

    for root, _, files in os.walk(directory):
        for name in files:
            path = os.path.join(root, name)
            try:
                file_hash = get_file_hash(path)
                hashes[file_hash].append(path)
            except (PermissionError, OSError):
                pass

    return {h: paths for h, paths in hashes.items() if len(paths) > 1}

if __name__ == "__main__":
    folder = input("Enter directory path: ")
    duplicates = find_duplicates(folder)

    if not duplicates:
        print("No duplicate files found.")
    else:
        print("Duplicate files:")
        for files in duplicates.values():
            for file in files:
                print(file)
            print("-" * 40)
